integration-tests/tests/src/tests/queries.ts

```diff
@@ -15,129 +15,1037 @@
-    contacts = realm.objects<IContact>(ContactSchema.name);
-    primitives = realm.objects<IPrimitive>(PrimitiveSchema.name);
+type QueryLengthPair = [ExpectedLength: number, Query: string, ...QueryArgs: Array<any>];
+type QueryExceptionPair = [ExpectedException: string, Query: string, ...QueryArgs: Array<any>];
+type QueryResultsPair = [ExpectedResults: any[], Query: string, ...QueryArgs: Array<any>];
+
+describe("Queries", () => {
+  /**
+   * Helper method which runs an array of queries and asserts them to given expected length
+   * For example: (r, "Obj", [1, "intCol == 0"]) => querying "intCol == 0" should return 1 elements.
+   */
+  const assertQueryLength = (
+    realm: Realm,
+    objectSchema: string | Realm.ObjectClass,
+    queryLengthPairs: QueryLengthPair[],
+  ) => {
+    queryLengthPairs.forEach(([expectedLength, queryString, ...queryArgs]) => {
+      const filtered = realm.objects(objectSchema).filtered(queryString, ...queryArgs);
+      expect(filtered.length).equal(
+        expectedLength,
+        `Expected length ${expectedLength} for query: ${queryString} ${JSON.stringify(queryArgs)}`,
+      );
+    });
+  };
+
+  /**
+   * Helper method which runs an array of queries and asserts them to throw given exception.
+   * For example: (r, "Obj", ["invalid", "intCol == #"]) => querying "intCol == #" should
+   * throw an exception which includes the string "invalid".
+   */
+  const assertQueryException = (
+    realm: Realm,
+    objectSchema: string | Realm.ObjectClass,
+    queryExceptionPairs: QueryExceptionPair[],
+  ) => {
+    queryExceptionPairs.forEach(([expectedException, queryString, ...queryArgs]) => {
+      expect(() => {
+        realm.objects(objectSchema).filtered(queryString, ...queryArgs);
+      }).throws(
+        expectedException,
+        `Expected exception not thrown for query: ${queryString} ${JSON.stringify(queryArgs)}`,
+      );
+    });
+  };
+
+  /**
+   * Helper method which runs an array of queries and asserts the results' provided [propertyToCompare]
+   * field to equal the given expected result array's values.
+   * For example: (r, "Obj", "intCol" [[3, 4], "intCol > 2]) => querying "intCol > 2" should
+   * return results whose elements' intCol property values are 3 and 4.
+   */
+  const assertQueryResultValues = (
+    realm: Realm,
+    objectSchema: string | Realm.ObjectClass,
+    propertyToCompare: string,
+    queryResultPairs: QueryResultsPair[],
+  ) => {
+    queryResultPairs.forEach(([expectedResults, queryString, ...queryArgs]) => {
+      let results = realm.objects<any>(objectSchema);
+      results = results.filtered(queryString, ...queryArgs);
+      expect(expectedResults.length).equals(results.length);
+      expect(expectedResults).to.deep.equal(
+        results.map((el) => el[propertyToCompare]),
+        `
+        Expected results not returned from query: ${queryString} ${JSON.stringify(queryArgs)}, 
+      `,
+      );
+    });
+  };
```

Comment:
Probably a matter of taste, but I'd suggest moving these out above of the `describe("Queries", () => {` and make then regular functions, instead of arrow functions (since they're not capturing a `this` of a parent scope anyway).

I also have a personal preference towards a for-of loop instead of calling `forEach` - we can talk more about this face-2-face, but basically, you're relying on the runtime implementation to call the callback synchronously - which it does on all runtimes I've seen, but then again ...

Reply:
That makes sense, will consider this thank you. With regards to arrow function vs regular functions, do you see regular functions as a default to use unless you'd benefit from not introducing another this into the scope? I think I have been looking at it from the opposite perspective (only use regular function if you need a local `this` scope).